### PR TITLE
Allow to use agent as init process and create initrd image based on rootfs

### PR DESCRIPTION
**Pull request by bergwolf**
The kata agent can be used as init process and thus we do not always rely on systemd. Also created kata initrd image based on rootfs created by `rootfs.sh`.

fixes: #5 
fixes: #7

### CODE REVIEW

File: initrd-builder/initrd_builder.sh [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+{
+	error="${1:-0}"
+	cat <<EOT
+Usage: ${SCRIPT_NAME} [options] <rootfs-dir>
+	This script will create a Kata Containers initrd image file based on the
+	<rootfs-dir> directory.
+
+Options:
+	-h Show this help
+	-o path to generate image file ENV: INITRD_IMAGE
```
Comment (jodh-intel) [inline initrd-builder/initrd_builder.sh]: Nit: You could also include the default value of `$INITRD_IMAGE` here.

File: initrd-builder/initrd_builder.sh [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+Extra environment variables:
+	USE_DOCKER: If set will build image in a Docker Container (requries docker)
```
Comment (jodh-intel) [inline initrd-builder/initrd_builder.sh]: Nit: "requires" (but I see that is also in the rootfs-builder and image-builder scripts too ;)

**Comment by jodh-intel** [comment]
Very nice PR! Apart from a couple of tiny nits, could you do a small update to https://github.com/kata-containers/osbuilder/tree/master/rootfs-builder#rootfs-requirements for `AGENT_INIT`? Once that's done,

lgtm

@klynnrif, @rcaballeromx - could you take a look at the doc changes please?

File: initrd-builder/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Creating a guest OS initrd image
+
+To create a guest OS initrd image run:
+
+```
+$ sudo ./initrd_builder.sh path/to/rootfs
+```
+
+Where `path/to/rootfs` is the directory populated by `rootfs.sh`.
```
Comment (rcaballeromx) [inline initrd-builder/README.md]: Convoluted, consider changing to:
The `rootfs.sh` script populates the `path/to/rootfs` directory.

File: initrd-builder/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+## Further information
+
+For more information about usage, run:
```
Comment (rcaballeromx) [inline initrd-builder/README.md]: Unspecific, consider changing to:
For more information on how to use the `initrd_builder.sh` script, run:

File: initrd-builder/initrd_builder.sh [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	local msg="$*"
+	echo "INFO: ${msg}"
+}
+
+usage()
+{
+	error="${1:-0}"
+	cat <<EOT
+Usage: ${SCRIPT_NAME} [options] <rootfs-dir>
+	This script will create a Kata Containers initrd image file based on the
```
Comment (rcaballeromx) [inline initrd-builder/initrd_builder.sh]: Avoid future tense, consider changing to:
This script creates a Kata Containers ...

File: initrd-builder/initrd_builder.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	<rootfs-dir> directory.
+
+Options:
+	-h Show this help
```
Comment (rcaballeromx) [inline initrd-builder/initrd_builder.sh]: Unnecessary 'this' and a missing period at the end.

File: initrd-builder/initrd_builder.sh (hunk 1 above)
Comment (rcaballeromx) [reply initrd-builder/initrd_builder.sh]: Missing verb, what does the -o option do? Does it set the output path, change it, or store it? I assume it sets the path. Consider changing to:
-o Set the path where the generated image file is stored. Default: the path stored in the environment variable INITRD_IMAGE

File: initrd-builder/initrd_builder.sh (hunk 2 above)
Comment (rcaballeromx) [reply initrd-builder/initrd_builder.sh]: Consider changing to:
USE_DOCKER: If set, the image builds in a Docker Container.  Setting this variable requires Docker.

File: rootfs-builder/rootfs.sh [hunk 7]
```diff
@@ -103,6 +107,14 @@ ENV PATH=\$PATH:\$GOROOT/bin:\$GOPATH/bin
 	popd
 }
 
+setup_agent_init() {
+	agent_bin="$1"
+	init_bin="$2"
+	info "Install $agent_bin as init process"
+	mv -f "${agent_bin}" ${init_bin}
```
Comment (jcvenegas) [inline rootfs-builder/rootfs.sh]: What about make a symlink from /sbin/init to the agent installed in bin?

Comment (bergwolf) [reply rootfs-builder/rootfs.sh]: When kata agent is used as the init process, there is no need for `/sbin/agent` to be present.

**Comment by jodh-intel** [comment]
Travis is failing in the `AGENT_INIT=yes` case:

```
# [OK] init is installed
# ERROR: /bin/kata-agent is not installed in /rootfs
# 	use AGENT_BIN env variable to change the expected agent binary name
```

**Comment by jodh-intel** [comment]
Hi @rcaballeromx, @jcvenegas - could you re-review please?

File: rootfs-builder/rootfs.sh (hunk 7 above)
Comment (jcvenegas) [reply rootfs-builder/rootfs.sh]: Yes, but doing this we avoid have an special case when agent is `init` for image/initrd building. Both both approaches sounds good.

File: initrd-builder/initrd_builder.sh [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+ROOTFS=$(readlink -f ${ROOTFS})
+IMAGE_DIR=$(dirname ${INITRD_IMAGE})
+IMAGE_DIR=$(readlink -f ${IMAGE_DIR})
+IMAGE_NAME=$(basename ${INITRD_IMAGE})
+
+# The kata rootfs image expects init to be installed
+init="${ROOTFS}/sbin/init"
+[ -x "${init}" ] || [ -L ${init} ] || die "/sbin/init is not installed in ${ROOTFS_DIR}"
+OK "init is installed"
```
Comment (jcvenegas) [inline initrd-builder/initrd_builder.sh]: here is not need an extra case when check if /bin/kata-agent is not installed if is not used?

**Comment by jodh-intel** [comment]
Hi @bergwolf - I've just realised that it would be very helpful if you added an `initd` term to:

- https://github.com/kata-containers/osbuilder#terms

... with a link to the new `initrd-builder/README.md` doc like the other terms.

File: rootfs-builder/rootfs.sh (hunk 7 above)
Comment (bergwolf) [reply rootfs-builder/rootfs.sh]: Actually the extra case helped me catch several places where `AGENT_INIT` env is not properly passed around. I hope to keep it to help future change as well.

File: initrd-builder/initrd_builder.sh (hunk 8 above)
Comment (bergwolf) [reply initrd-builder/initrd_builder.sh]: Good point! I need to check `/usr/bin/kata-agent` presence if `AGENT_INIT` is no.

**Comment by bergwolf** [comment]
@jodh-intel @jcvenegas PR updated. PTAL.

File: initrd-builder/initrd_builder.sh [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+info()
+{
+	local msg="$*"
+	echo "INFO: ${msg}"
+}
+
+usage()
+{
+	error="${1:-0}"
```
Comment (jodh-intel) [inline initrd-builder/initrd_builder.sh]: That is neat! :smile:

File: README.md [hunk 10]
```diff
@@ -30,6 +30,14 @@ This section describes the terms used for all documentation in this repository.
 
   See [the image builder documentation](image-builder/README.md).
 
+- initrd (or "initramfs")
```
Comment (jodh-intel) [inline README.md]: Thanks!

**Comment by jodh-intel** [comment]
lgtm

**Comment by jodh-intel** [comment]
@rcaballeromx, @klynnrif - could you take a look at the doc changes please?

**Comment by jodh-intel** [comment]
@rcaballeromx, @klynnrif - I think the only un-reviewed doc changes are the recent changes to the top-level `README.md` (new "initrd" section).

File: README.md [hunk 11]
```diff
@@ -30,6 +30,14 @@ This section describes the terms used for all documentation in this repository.
...
+
+  A compressed cpio archive loaded into memory and used as part of the Linux
+  startup process. During startup, it is unpacked by the kernel into a special
```
Comment (rcaballeromx) [inline README.md]: Passive voice. Consider changing to:
During startup, the kernel unpacks it into a special

File: rootfs-builder/README.md [hunk 12]
```diff
@@ -38,6 +39,8 @@ The rootfs must provide at least the following components:
 
   Path: `/sbin/init` - init binary called by the kernel.
 
+When `AGENT_INIT` environment variable is set to `yes`, use Kata agent as `/sbin/init`.
```
Comment (rcaballeromx) [inline rootfs-builder/README.md]: Missing article:
When the `AGENT_INIT` environment variable ...

File: rootfs-builder/README.md [hunk 13]
```diff
@@ -46,6 +49,15 @@ To build a rootfs for your chosen distribution, run:
...
 ```
 
+## Creating a rootfs with kernel modules
+
+To build a rootfs with additional kernel modules, run:
+```
+$ sudo KERNEL_MODULES_DIR=${kernel_mod_dir} ./rootfs.sh <distro>
+```
+Where `kernel_mod_dir` points to the kernel modules directory to be put under
+`/lib/modules/` directory of the created rootfs.
```
Comment (rcaballeromx) [inline rootfs-builder/README.md]: Missing article:
the `/lib/modules/` directory of the created rootfs.

**Comment by bergwolf** [comment]
@rcaballeromx thanks! updated.

**Comment by jodh-intel** [comment]
@sameo, @jcvenegas - could you maybe take a look please?

**Comment by sameo** [comment]
LGTM

**Pull request merged**